linker/shared/src/main/scala/org/scalajs/linker/frontend/modulesplitter/MaxModuleAnalyzer.scala

```diff
@@ -175,51 +176,73 @@ private object MaxModuleAnalyzer {
     }
   }
 
-  /** Set of shortest, mutually prefix-free paths. */
-  private final class Paths[H, T] private (
-      private val content: mutable.Map[H, Paths[T, T]]) {
+  /** "Interesting" paths that can lead to a given class.
+   *
+   *  "Interesting" in this context means:
+   *  - All direct paths from a public dependency.
+   *  - All non-empty, mutually prefix free paths, of dynamic import hops.
+   */
+  private final class Paths {
+    private val direct = mutable.Set.empty[ModuleID]
+    private val dynamic = mutable.Map.empty[ModuleID, DynamicPaths]
 
-    /* cannot make this tailrec, because the type parameters change over the
-     * recursion. 2.11 does not support this.
-     */
-    def put(h: H, t: List[T]): Boolean = {
-      if (content.get(h).exists(_.isEmpty)) {
+    def put(pathRoot: ModuleID, pathSteps: List[ClassName]): Boolean = {
+      if (pathSteps.isEmpty) {
+        direct.add(pathRoot)
+      } else {
+        dynamic
+          .getOrElseUpdate(pathRoot, new DynamicPaths)
+          .put(pathSteps)
+      }
+    }
+
+    def tags: (Set[ModuleID], Set[ClassName]) = {
+      /* Remove dynamic paths to class that are also reached by a public module.
+       * However, only do this if there are other tags as well. Otherwise, this
+       * class will end up in a public module, but the dynamically loaded module
+       * will try to import it (but importing public modules is forbidden).
+       */
+      if (direct.size > 1 || direct != dynamic.keySet) {
+        direct.foreach(dynamic.remove(_))
```

Comment:
Since there are side-effects in this method, `tags` should be declared and called with `()`, so as `def tags()`.